### DISCUSSION TITLE
Run program through USB <pybricksdev>

### DISCUSSION THREAD

**Discussion by edwardhura** [pybricks/support#1548] [Q&A] [answered]
Hi! I have start working with pybriks and tried to figured out how I can run programs through USB. I have found few threads where were some disscussion about this ability, but I still not understand yet.
Is it possible to run programs through USB??

I have found this in readme for pybriksdev repo:
```
#
# usb connection examples:
# NOTE: running programs via usb connection works for official LEGO firmwares only

# Run script on any Pybricks device
pipx run pybricksdev run usb demo/shortdemo.py
```

And it confusing me ^. What I understood: run through USB is available for Lego Firmwares only and just right after I see text that shows me how to run programs on Pybricks firmware devices... Am I wrong?

**Comment by laurensvalk** [comment]
Some work is ongoing to make this possible, but it isn't finished yet.

What is your use case for using USB?

**Reply by edwardhura** [reply]
Ability to work with XBox Controller and debug without disconnecting (I could be wrong because I haven't done it yet, just read different threads, my xbox controller will be delivered only tomorrow).
 But I agree on that BLE connection is more comfortable for usage in case when u have builded some device.

Still not clear this  ` NOTE: running programs via usb connection works for official LEGO firmwares only` - does it mean that I can run mindstorm programs with pybriksdev through USB connection?

**Reply by laurensvalk** [reply, marked as answer]
> Ability to work with XBox Controller and debug without disconnecting

Ah - fortunately, this has nothing to do with USB :smile: You can just use the Mindstorms/Spike Hub with Bluetooth, then :)

The disconnecting issue only appears on Technic Hub.

> Still not clear this  NOTE: running programs via usb connection works for official LEGO firmwares only - does it mean that I can run mindstorm programs with pybriksdev through USB connection?

Some work is ongoing to make USB support possible, but it isn't finished yet.

Pybricksdev is mainly just an internal development tool. Experienced users may get some value out of it too, but we don't actively support this.

**Reply by edwardhura** [reply]
Got you! Thanks for fast response in my question. That was helpful!

**Reply by laurensvalk** [reply]
Have fun with the Xbox controller when it arrives!

**Comment by benjaminrovner** [comment]
Is it possible to download code using USB to a hub running pybricks?

**Reply by dlech** [reply]
Yes, but it is still experimental at this point. You can try it at https://labs.pybricks.com/.

And for SPIKE Prime, you will need to do a custom firmware build with USB enabled.

**Reply by zoligyenge** [reply]
> Yes, but it is still experimental at this point. You can try it at https://labs.pybricks.com/.
> 
> And for SPIKE Prime, you will need to do a custom firmware build with USB enabled.

Hi. Is there any difference between https://**labs.**pybricks.com and https://**beta.**pybricks.com/?

**Reply by dlech** [reply]
`labs` can be updated at any time for any reason - it is very experimental. `beta` is for testing the next release. Sometimes these may happen to have the same version, but usually not.
